### ISSUE TITLE
C++ boost intellisense how to get it working?

### ISSUE THREAD

**Issue by JesseZhuang** [microsoft/vscode-cpptools#10315]
### Environment

- OS and Version: Mac Montery 12.6.1
- VS Code Version: 1.74.2
- C/C++ Extension Version: 1.13.8

Has anyone got intellisense working for c++ boost.test with Visual Studio code?

boost is installed in `/usr/local/Cellar/boost/1.80.0/include`

Things have tried:

1. add the above path to c++ intellisense configuration as `/usr/local/Cellar/boost/1.80.0/include/**` or `/usr/local/Cellar/boost/1.80.0/include`
1. add `-L /usr/local/Cellar/boost/1.80.0/include` to compiler args in to c++ intellisense configuration
1. add `-L /usr/local/Cellar/boost/1.80.0/include` to compiler args in to tasks.json
1. add path to settings.json as browse path and/or include path

Not working so far.

Screenshot: https://i.stack.imgur.com/LEIJT.png

### Bug Summary and Steps to Reproduce

Bug Summary:

Intellisense includepath not working for c++ boost library

Steps to reproduce:
1. install boost
2. add boot library path to includepath
3. vs code still not able to find boost unit test header file, verified the hpp file is there

Expected behavior: vs code able to find the file and auto complete working

### Configuration and Logs

c_cpp_properties.json

```shell
{
    "configurations": [
        {
            "name": "Mac",
            "includePath": [
                "${workspaceFolder}/**",
                "/usr/local/Cellar/boost/1.80.0/include/**"
            ],
            "defines": [],
            "macFrameworkPath": [
                "/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/System/Library/Frameworks"
            ],
            "compilerPath": "/usr/bin/clang",
            "cStandard": "c17",
            "cppStandard": "c++17",
            "intelliSenseMode": "macos-clang-x64",
            "configurationProvider": "ms-vscode.cmake-tools",
            "compilerArgs": [
                "-L /usr/local/Cellar/boost/1.80.0/include"
            ]
        }
    ],
    "version": 4
}
```

**Comment by sean-mcmanus** [comment]
You have `"configurationProvider": "ms-vscode.cmake-tools",` set so you need to configure CMake (Tools) to use boost and it should send that info to us. If you run the C/C++: Log Diagnostics command after opening a file you can see what configuration info the CMake Tools extension is sending us.

However, you can also set `"/usr/local/Cellar/boost/1.80.0/include"` in the includePath and also set             `"mergeConfigurations": true,` and that should merge the two. Don't set it with "**" at the end and don't set it in the compiler args.

**Comment by JesseZhuang** [comment]
Diagnostics without the boost path in `c_cpp_properties.json`, intellisense not working

```shell
-------- Diagnostics - 12/28/2022, 6:40:48 PM
Version: 1.13.8
Current Configuration:
{
    "name": "Mac",
    "includePath": [
        "${workspaceFolder}/**"
    ],
    "defines": [],
    "macFrameworkPath": [
        "/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/System/Library/Frameworks"
    ],
    "compilerPath": "/usr/bin/clang",
    "cStandard": "c17",
    "cppStandard": "c++17",
    "intelliSenseMode": "macos-clang-x64",
    "configurationProvider": "ms-vscode.cmake-tools",
    "compilerPathIsExplicit": true,
    "cStandardIsExplicit": true,
    "cppStandardIsExplicit": true,
    "intelliSenseModeIsExplicit": true,
    "mergeConfigurations": false,
    "browse": {
        "path": [],
        "limitSymbolsToIncludedHeaders": true
    }
}
Custom browse configuration: 
{
    "browsePath": [
        "/volumes/workplace/projects/csapp/build/cmakefiles",
        "/volumes/workplace/projects/csapp/cpp"
    ],
    "compilerPath": "/usr/bin/clang++",
    "compilerArgs": [],
    "compilerFragments": [
        "-g -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX13.1.sdk -mmacosx-version-min=12.6   -std=c++17"
    ]
}
Translation Unit Mappings:
[ /Volumes/workplace/projects/CSAPP/cpp/leetcode/test/test_foo.cpp ]:
    /Volumes/workplace/projects/CSAPP/cpp/leetcode/test/test_foo.cpp
Translation Unit Configurations:
[ /Volumes/workplace/projects/CSAPP/cpp/leetcode/test/test_foo.cpp ]:
    Process ID: 19730
    Memory Usage: 9 MB
    Compiler Path: /usr/bin/clang++
    Includes:
        /Library/Developer/CommandLineTools/SDKs/MacOSX13.1.sdk/usr/include/c++/v1
        /Library/Developer/CommandLineTools/usr/lib/clang/14.0.0/include
        /Library/Developer/CommandLineTools/SDKs/MacOSX13.1.sdk/usr/include
        /Library/Developer/CommandLineTools/usr/include
    Frameworks:
        /Library/Developer/CommandLineTools/SDKs/MacOSX13.1.sdk/System/Library/Frameworks
    Standard Version: c++17
    IntelliSense Mode: macos-clang-x64
    Other Flags:
        --clang
        --clang_version=110000
Total Memory Usage: 9 MB

------- Workspace parsing diagnostics -------
Number of files discovered (not excluded): 2633
```

**Comment by JesseZhuang** [comment]
Tried your suggestion with following `c_cpp_properties.json`, still not working

```
{
    "configurations": [
        {
            "name": "Mac",
            "includePath": [
                "${workspaceFolder}/**",
                "/usr/local/Cellar/boost/1.80.0/include"
            ],
            "defines": [],
            "macFrameworkPath": [
                "/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/System/Library/Frameworks"
            ],
            "compilerPath": "/usr/bin/clang",
            "cStandard": "c17",
            "cppStandard": "c++17",
            "intelliSenseMode": "macos-clang-x64",
            "configurationProvider": "ms-vscode.cmake-tools",
            "mergeConfigurations": true
        }
    ],
    "version": 4
}
```

Diagnostics below

```
-------- Diagnostics - 12/28/2022, 6:56:40 PM
Version: 1.13.8
Current Configuration:
{
    "name": "Mac",
    "includePath": [
        "${workspaceFolder}/**",
        "/usr/local/Cellar/boost/1.80.0/include"
    ],
    "defines": [],
    "macFrameworkPath": [
        "/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/System/Library/Frameworks"
    ],
    "compilerPath": "/usr/bin/clang",
    "cStandard": "c17",
    "cppStandard": "c++17",
    "intelliSenseMode": "macos-clang-x64",
    "configurationProvider": "ms-vscode.cmake-tools",
    "mergeConfigurations": true,
    "compilerPathIsExplicit": true,
    "cStandardIsExplicit": true,
    "cppStandardIsExplicit": true,
    "intelliSenseModeIsExplicit": true,
    "browse": {
        "path": [],
        "limitSymbolsToIncludedHeaders": true
    }
}
Custom browse configuration: 
{
    "browsePath": [
        "/volumes/workplace/projects/csapp/build/cmakefiles",
        "/volumes/workplace/projects/csapp/cpp"
    ],
    "compilerPath": "/usr/bin/clang++",
    "compilerArgs": [],
    "compilerFragments": [
        "-g -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX13.1.sdk -mmacosx-version-min=12.6   -std=c++17"
    ]
}
Translation Unit Mappings:
[ /Volumes/workplace/projects/CSAPP/cpp/leetcode/test/test_foo.cpp ]:
    /Volumes/workplace/projects/CSAPP/cpp/leetcode/test/test_foo.cpp
Translation Unit Configurations:
[ /Volumes/workplace/projects/CSAPP/cpp/leetcode/test/test_foo.cpp ]:
    Process ID: 21182
    Memory Usage: 8 MB
    Compiler Path: /usr/bin/clang++
    Includes:
        /Library/Developer/CommandLineTools/SDKs/MacOSX13.1.sdk/usr/include/c++/v1
        /Library/Developer/CommandLineTools/usr/lib/clang/14.0.0/include
        /Library/Developer/CommandLineTools/SDKs/MacOSX13.1.sdk/usr/include
        /Library/Developer/CommandLineTools/usr/include
    Frameworks:
        /Library/Developer/CommandLineTools/SDKs/MacOSX13.1.sdk/System/Library/Frameworks
    Standard Version: c++17
    IntelliSense Mode: macos-clang-x64
    Other Flags:
        --clang
        --clang_version=110000
Total Memory Usage: 8 MB

------- Workspace parsing diagnostics -------
Number of files discovered (not excluded): 2633


```

**Comment by sean-mcmanus** [comment]
What version of CMake Tools are you using? You should use 1.13.31 or older due to a regression, i.e. I repro this bug with 1.13.32 or 1.13.33, but otherwise I don't. Also, were it to work you'd want to remove the `"${workspaceFolder}/**"`. The code that handles this is at https://github.com/microsoft/vscode-cpptools/blob/main/Extension/src/LanguageServer/client.ts#L1643 , i.e. from reviewing that code it looks like mergeConfigurations will fail to work as expected if the configuration provider fails "too early" and doesn't actually provide configurations.

**Comment by JesseZhuang** [comment]
Thanks for filing the issue with cpptools.

Cmake version is 3.25.1 are you suggesting version 1.13.31 will work?

**Comment by sean-mcmanus** [comment]
@JesseZhuang I'm referring to the 1.13.31 of CMake Tools, not the CMake extension. It may fix the issue if you're using 1.13.33.

**Comment by JesseZhuang** [comment]
I had these installed:

1. cmake installed with homebrew: version 3.25.1
2. boost installed with homebrew: version 1.18.0
3. c++ extension pack in vs code: https://marketplace.visualstudio.com/items?itemName=ms-vscode.cpptools-extension-pack version 1.3.0, which includes cmake extension version 0.0.17 and cmake tools version 1.12.17

I tried switching to version 1.13.31 (pre-release) for the cmake tools extension and tried removing the ""${workspaceFolder}/**" but still did not get intellisense working.
